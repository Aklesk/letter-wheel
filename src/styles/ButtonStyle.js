import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  wrapper: {
    width: '100%',
  },
  image: {
    ...StyleSheet.absoluteFillObject,
    alignItems: 'center',
    justifyContent: 'center',
    width: undefined,
    height: undefined,
    marginRight: -4,
  },
  textWrapper: {
    alignItems: 'center',
    justifyContent: 'center',
    paddingBottom: 2,
    paddingLeft: 12,
    paddingRight: 12,
    paddingTop: 8,
  },
  text: {
    fontFamily: 'Grundschrift-Bold',
    color: '#000000',
    fontSize: 20,
    marginBottom: 15,
  },
});
