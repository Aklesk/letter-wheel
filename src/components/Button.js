import React from 'react';
import PropTypes from 'prop-types';
import { Image, Text, TouchableWithoutFeedback, View } from 'react-native';
import { observer } from 'mobx-react';
import { action, observable } from 'mobx';

import ButtonStyle from '../styles/ButtonStyle';

import buttonYellowPassive from '../images/buttonYellowPassive.png';
import buttonYellowActive from '../images/buttonYellowActive.png';
import buttonBluePassive from '../images/buttonBluePassive.png';
import buttonBlueActive from '../images/buttonBlueActive.png';

@observer
export default class Button extends React.Component {
  static propTypes = {
    colour: PropTypes.oneOf(['yellow', 'blue']),
    title: PropTypes.string,
    onPress: PropTypes.func,
  }
  static defaultProps = {
    colour: 'yellow',
    title: '',
    onPress() {},
  }

  @action beginPress = () => {
    this.isPressing = true;
  }
  @action endPress = () => {
    this.isPressing = false;
  }
  @observable isPressing = false;

  render() {
    let unpressed;
    let pressed;

    if (this.props.colour === 'yellow') {
      unpressed = buttonYellowPassive;
      pressed = buttonYellowActive;
    }
    if (this.props.colour === 'blue') {
      unpressed = buttonBluePassive;
      pressed = buttonBlueActive;
    }

    return (
      <View style={ButtonStyle.wrapper}>
        <Image
          fadeDuration={0}
          resizeMode="stretch"
          source={pressed}
          style={[
            ButtonStyle.image,
            { opacity: this.isPressing ? 1 : 0 },
          ]}
        />
        <Image
          fadeDuration={0}
          resizeMode="stretch"
          source={unpressed}
          style={[
            ButtonStyle.image,
            { opacity: this.isPressing ? 0 : 1 },
          ]}
        />
        <TouchableWithoutFeedback
          onPress={this.props.onPress}
          onPressIn={this.beginPress}
          onPressOut={this.endPress}
        >
          <View style={ButtonStyle.textWrapper}>
            <Text
              style={ButtonStyle.text}
            >
              {this.props.title.toUpperCase()}
            </Text>
          </View>
        </TouchableWithoutFeedback>
      </View>
    );
  }
}
