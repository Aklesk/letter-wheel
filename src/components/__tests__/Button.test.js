import React from 'react';
import { Image, TouchableWithoutFeedback } from 'react-native';
import { mount } from 'enzyme';
import { isAction, isObservableProp, runInAction } from 'mobx';

import Button from '../Button';
import ButtonStyle from '../../styles/ButtonStyle';

describe('Button component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('matches the snapshot', () => {
    const render = mount((
      <Button />
    ));
    expect(render).toMatchSnapshot();
  });

  it('provides an \'isPressing\' observable', () => {
    const render = mount((
      <Button />
    ));
    expect(render.instance().isPressing).toEqual(false);
    expect(isObservableProp(render.instance(), 'isPressing')).toBeTruthy();
  });

  it('provides a \'beginPress\' action method', () => {
    const render = mount((
      <Button />
    ));
    expect(render.instance().beginPress).toEqual(expect.any(Function));
    expect(isAction(render.instance().beginPress)).toBeTruthy();
    expect(render.instance().isPressing).toEqual(false);
    expect(render.instance().beginPress()).toBeUndefined();
    expect(render.instance().isPressing).toEqual(true);
  });

  it('provides a \'endPress\' action method', () => {
    const render = mount((
      <Button />
    ));
    runInAction(() => { render.instance().isPressing = true; });
    expect(render.instance().endPress).toEqual(expect.any(Function));
    expect(isAction(render.instance().endPress)).toBeTruthy();
    expect(render.instance().isPressing).toEqual(true);
    expect(render.instance().endPress()).toBeUndefined();
    expect(render.instance().isPressing).toEqual(false);
  });

  it('renders a touchable that manages isPressing state and triggers onPress', () => {
    const onPress = jest.fn();
    const render = mount((
      <Button
        onPress={onPress}
      />
    ));
    const touchable = render.find(TouchableWithoutFeedback).at(0);
    expect(touchable.props().onPressIn).toEqual(render.instance().beginPress);
    expect(touchable.props().onPressOut).toEqual(render.instance().endPress);
    expect(touchable.props().onPress).toEqual(onPress);
  });

  it('renders correct images based on colour prop and press state', () => {
    const render = mount((
      <Button />
    ));
    expect(render.find(Image)).toHaveLength(2);
    expect(render.find(Image).at(0).props().source.testUri)
      .toEqual('src/images/buttonYellowActive.png');
    expect(render.find(Image).at(0).props().style).toEqual([ButtonStyle.image, { opacity: 0 }]);
    expect(render.find(Image).at(1).props().source.testUri)
      .toEqual('src/images/buttonYellowPassive.png');
    expect(render.find(Image).at(1).props().style).toEqual([ButtonStyle.image, { opacity: 1 }]);

    runInAction(() => { render.instance().isPressing = true; });
    render.update();

    expect(render.find(Image).at(0).props().style).toEqual([ButtonStyle.image, { opacity: 1 }]);
    expect(render.find(Image).at(1).props().style).toEqual([ButtonStyle.image, { opacity: 0 }]);


    runInAction(() => { render.instance().isPressing = false; });
    render.setProps({ colour: 'blue' });
    render.update();


    expect(render.find(Image)).toHaveLength(2);
    expect(render.find(Image).at(0).props().source.testUri)
      .toEqual('src/images/buttonBlueActive.png');
    expect(render.find(Image).at(0).props().style).toEqual([ButtonStyle.image, { opacity: 0 }]);
    expect(render.find(Image).at(1).props().source.testUri)
      .toEqual('src/images/buttonBluePassive.png');
    expect(render.find(Image).at(1).props().style).toEqual([ButtonStyle.image, { opacity: 1 }]);

    runInAction(() => { render.instance().isPressing = true; });
    render.update();

    expect(render.find(Image).at(0).props().style).toEqual([ButtonStyle.image, { opacity: 1 }]);
    expect(render.find(Image).at(1).props().style).toEqual([ButtonStyle.image, { opacity: 0 }]);
  });
});
