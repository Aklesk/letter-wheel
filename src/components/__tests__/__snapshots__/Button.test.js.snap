// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component matches the snapshot 1`] = `
<Button
  colour="yellow"
  onPress={[Function]}
  title=""
>
  <View
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <react-native-mock>
      <Image
        fadeDuration={0}
        resizeMode="stretch"
        source={
          Object {
            "testUri": "src/images/buttonYellowActive.png",
          }
        }
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": undefined,
              "justifyContent": "center",
              "marginRight": -4,
              "width": undefined,
            },
            Object {
              "opacity": 0,
            },
          ]
        }
      >
        <react-native-mock />
      </Image>
      <Image
        fadeDuration={0}
        resizeMode="stretch"
        source={
          Object {
            "testUri": "src/images/buttonYellowPassive.png",
          }
        }
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": undefined,
              "justifyContent": "center",
              "marginRight": -4,
              "width": undefined,
            },
            Object {
              "opacity": 1,
            },
          ]
        }
      >
        <react-native-mock />
      </Image>
      <TouchableWithoutFeedback
        onPress={[Function]}
        onPressIn={[Function]}
        onPressOut={[Function]}
      >
        <react-native-mock>
          <View
            style={
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "paddingBottom": 2,
                "paddingLeft": 12,
                "paddingRight": 12,
                "paddingTop": 8,
              }
            }
          >
            <react-native-mock>
              <Text
                style={
                  Object {
                    "color": "#000000",
                    "fontFamily": "Grundschrift-Bold",
                    "fontSize": 20,
                    "marginBottom": 15,
                  }
                }
              >
                <react-native-mock />
              </Text>
            </react-native-mock>
          </View>
        </react-native-mock>
      </TouchableWithoutFeedback>
    </react-native-mock>
  </View>
</Button>
`;
